### PR TITLE
Added DENY_ALL policy option

### PR DESCRIPTION
**Pull request by vmanilo**
Github Issue: https://github.com/Twingate/terraform-provider-twingate/issues/70

## Changes
- Added DENY_ALL policy option

### CODE REVIEW

**Comment by vmanilo** [comment]
Guys, what do you think? it is all that should be addressed in this PR regarding the issue?

From my point of view, provider is lacking policy validation, but it also depends how backend handles it if user provides request like this:

```
protocols {
    tcp  {
      policy = "ALLOW_ALL"
      ports = ["80", "82-83"]
    }
  }
```

**Comment by vmanilo** [comment]
Currently I see error like this when trying to use DENY_ALL policy type

```
│ Error: failed to create resource: non-200 OK status code: 400 Bad Request body: "{\"errors\":[{\"message\":\"Variable '$protocols' got invalid value 'DENY_ALL' at 'protocols.udp.policy'; Value 'DENY_ALL' does not exist in 'ProtocolPolicy' enum.\",\"locations\":[{\"line\":1,\"column\":69}],\"path\":null}]}"
```

because it's missing on graphQL side
Screenshot 2022-04-23 at 11 37 39

**Comment by vmanilo** [comment]
Should I disable test with deny_all policy for now, and re-enable when graphQL will be updated?

File: twingate/resource_resource_test.go [hunk 1]
```diff
@@ -112,6 +119,29 @@ func testTwingateResource_withProtocolsAndGroups(networkName, groupName1, groupN
...
+func testTwingateResource_withDenyAll(networkName, resourceName string) string {
+	return fmt.Sprintf(`
+	resource "twingate_remote_network" "test" {
+	  name = "%s"
+	}
+	resource "twingate_resource" "test" {
+	  name = "%s"
+	  address = "updated-acc-test.com"
+	  remote_network_id = twingate_remote_network.test.id
+	  group_ids = ["R3JvdXA6MjMxNTQ="]
```
Comment (vmanilo) [inline twingate/resource_resource_test.go:131]: will fix this with group resource

File: docs/resources/resource.md [hunk 2]
```diff
@@ -22,16 +22,20 @@ resource "twingate_remote_network" "aws_network" {
...
   remote_network_id = twingate_remote_network.aws_network.id
-  group_ids = ["group1"]
+  group_ids = [twingate_group.aws.id]
   protocols {
     allow_icmp = true
     tcp  {
-      policy = "RESTRICTED"
-      ports = ["80", "82-83"]
+      policy = "DENY_ALL"
+      ports = ["80", "82-83", "90"]
```
Comment (ekampf) [inline docs/resources/resource.md:38]: What does it mean when you `DENY_ALL` but also give `ports` ?
This should error... or warn?

File: examples/resources/twingate_resource/resource.tf [hunk 3]
```diff
@@ -7,16 +7,20 @@ resource "twingate_remote_network" "aws_network" {
...
   remote_network_id = twingate_remote_network.aws_network.id
-  group_ids = ["group1"]
+  group_ids = [twingate_group.aws.id]
   protocols {
     allow_icmp = true
     tcp  {
-      policy = "RESTRICTED"
-      ports = ["80", "82-83"]
+      policy = "DENY_ALL"
+      ports = ["80", "82-83", "90"]
```
Comment (ekampf) [inline examples/resources/twingate_resource/resource.tf:23]: same

File: docs/resources/resource.md (hunk 2 above)
Comment (vmanilo) [reply docs/resources/resource.md:38]: ports will be ignored in this case

Comment (vmanilo) [reply docs/resources/resource.md:38]: yeah, this was not specified

Comment (ekampf) [reply docs/resources/resource.md:38]: So we shouldn't give that as an example :). you should probably change it back to `RESTRICTED`

Comment (vmanilo) [reply docs/resources/resource.md:38]: ok, will change to `RESTRICTED`

and in case user provides non-empty list of ports with DENY_ALL, what error message should be?

Comment (vmanilo) [reply docs/resources/resource.md:38]: @ekampf changed back to `RESTRICTED`

File: examples/resources/twingate_resource/resource.tf (hunk 3 above)
Comment (vmanilo) [reply examples/resources/twingate_resource/resource.tf:23]: @ekampf changed back to `RESTRICTED`

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,7 @@
 ## 0.0.1 (Unreleased)
```
Comment (ekampf) [inline CHANGELOG.md:1]: This changelog is not up to date. We're already at `0.1.6` and this would release in `0.1.7`
Maybe we should just get rid of this file as we manage changelog in the releases screen

File: .go-version [hunk 5]
```diff
@@ -1 +1 @@
-1.16.2
+1.18.2
```
Comment (ekampf) [inline .go-version:1]: Why is upgrading Go version + dependencies part of these PRs?
Are these required to add "DENY_ALL" policy?   if not they should be separate PRs.

PRs should be as small as possible and limited to "single" change.

File: docs/resources/connector_tokens.md [hunk 6]
```diff
@@ -40,12 +40,12 @@ resource "twingate_connector_tokens" "aws_connector_tokens" {
 
 ### Optional
 
-- `id` (String) The ID of this resource.
 - `keepers` (Map of String) Arbitrary map of values that, when changed, will trigger recreation of resource. Use this to automatically rotate Connector tokens on a schedule.
 
 ### Read-Only
 
 - `access_token` (String, Sensitive) The Access Token of the parent Connector
+- `id` (String) The ID of this resource.
```
Comment (ekampf) [inline docs/resources/connector_tokens.md:48]: `id` should be first. Then `access_token` and `refresh_token`

**Review by ekampf — CHANGES_REQUESTED** [review]
This PR seems to include multiple separate changes.
There are at least 3 things I see here that should be separate PRs:
- Upgrade Go and dependencies
- Upgrade documentation\examples to use `group` resources instead of hardcode
- Add DENY_ALL option to resource policy

Can we please split these into small and focused PRs?

File: CHANGELOG.md (hunk 4 above)
Comment (ekampf) [reply CHANGELOG.md:1]: Made a separate PR for this - https://github.com/Twingate/terraform-provider-twingate/pull/88

File: examples/resources/twingate_resource/import.sh [hunk 7]
```diff
@@ -0,0 +1,3 @@
+terraform import twingate_remote_network.network UmVtb3RlTmV0d29yazo0MDEzOQ==
```
[the remark this answers is not in the document]
Comment (vmanilo) [reply examples/resources/twingate_resource/import.sh:1]: import script, the same we have for other resources as well

it's not necessary, just allows to fetch existing resource

File: docs/resources/connector_tokens.md (hunk 6 above)
Comment (vmanilo) [reply docs/resources/connector_tokens.md:48]: ok

File: examples/resources/twingate_resource/import.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (vmanilo) [reply examples/resources/twingate_resource/import.sh:1]: just as example of importing existing resources

**Comment by vmanilo** [comment]
> This PR seems to include multiple separate changes. There are at least 3 things I see here that should be separate PRs:
> 
> * Upgrade Go and dependencies
> * Upgrade documentation\examples to use `group` resources instead of hardcode
> * Add DENY_ALL option to resource policy
> 
> Can we please split these into small and focused PRs?

Separated changes in different PRs:
* https://github.com/Twingate/terraform-provider-twingate/pull/89
* https://github.com/Twingate/terraform-provider-twingate/pull/91
* https://github.com/Twingate/terraform-provider-twingate/pull/90

File: twingate/resource_resource.go [hunk 8]
```diff
@@ -11,14 +11,33 @@ import (
...
+	policyAllowAll   = "ALLOW_ALL"
+	policyDenyAll    = "DENY_ALL"
+)
+
+func castToStrings(a, b interface{}) (string, string) {
+	return a.(string), b.(string)
+}
+
+func protocolDiff(k, oldValue, newValue string, d *schema.ResourceData) bool {
+	oldPolicy, newPolicy := castToStrings(d.GetChange("protocols.0.tcp.0.policy"))
```
Comment (vmanilo) [inline twingate/resource_resource.go:25]: need to double-check, why here is only `tcp` protocol, seems like I didn't cover `udp` case

Comment (vmanilo) [reply twingate/resource_resource.go]: fixed

File: examples/resources/twingate_resource/import.sh (hunk 7 above)
[the remark this answers is not in the document]
Comment (alexmensch) [reply examples/resources/twingate_resource/import.sh:1]: This is nice, glad you added this example

**Comment by alexmensch** [comment]
@romankor could you review this please?

**Review by ekampf — APPROVED** [review]

**Comment by ekampf** [comment]
@vmanilo please fix documentation comments before merging

File: examples/resources/twingate_resource/import.sh [hunk 9]
```diff
@@ -0,0 +1,3 @@
...
+terraform import twingate_group.aws R3JvdXA6MzUxNjU=
+terraform import twingate_resource.resource UmVzb3VyY2U6MzQwNDQ3
```
[the remark this answers is not in the document]
Comment (vmanilo) [reply examples/resources/twingate_resource/import.sh:1]: fixed
